pkg/gadgets/ebpftop/piditer/iter.go

```diff
@@ -0,0 +1,191 @@
+// Copyright 2019-2022 The Inspektor Gadget authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package piditer
+
+import (
+	"bufio"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"strconv"
+	"strings"
+	"sync"
+	"unsafe"
+
+	"golang.org/x/sys/unix"
+
+	"github.com/cilium/ebpf"
+	"github.com/cilium/ebpf/link"
+)
+
+//go:generate go run github.com/cilium/ebpf/cmd/bpf2go -target $TARGET -cc clang -type pid_iter_entry piditer ./bpf/pid_iter.bpf.c -- -I./bpf/ -I../../../${TARGET}
+
+type pidIter struct {
+	objs piditerObjects
+	iter *link.Iter
+}
+
+type PidIterEntry struct {
+	ProgID uint32
+	Pid    uint32
+	Comm   string
+}
+
+var (
+	addrLock        sync.Mutex
+	bpfProgFopsAddr uint64
+	triedGetAddr    bool
+
+	iterEntrySize = int(unsafe.Sizeof(piditerPidIterEntry{}))
+)
+
+// getBpfProgFopsAddr reads the address of bpf_prog_fops from /proc/kallsyms. It
+// only does this on first call and caches the address.
+func getBpfProgFopsAddr() uint64 {
+	addrLock.Lock()
+	defer addrLock.Unlock()
+
+	if triedGetAddr {
+		return bpfProgFopsAddr
+	}
+
+	triedGetAddr = true
+
+	file, err := os.Open("/proc/kallsyms")
+	if err != nil {
+		return 0
+	}
+	defer file.Close()
+
+	scanner := bufio.NewScanner(file)
+	for scanner.Scan() {
+		line := scanner.Text()
+
+		fields := strings.Fields(line)
+		if len(fields) < 3 {
+			continue
+		}
+
+		if fields[2] == "bpf_prog_fops" {
+			addr, err := strconv.ParseUint(fields[0], 16, 64)
+			if err != nil {
+				return 0
+			}
+
+			// cache addr
+			bpfProgFopsAddr = addr
+			return bpfProgFopsAddr
+		}
+	}
+	return bpfProgFopsAddr
+}
+
+func NewPidIter() (*pidIter, error) {
+	spec, err := loadPiditer()
+	if err != nil {
+		return nil, fmt.Errorf("failed to load ebpf program: %w", err)
+	}
+
+	consts := map[string]interface{}{
+		"bpf_prog_fops_addr": getBpfProgFopsAddr(),
+	}
+
+	if err := spec.RewriteConstants(consts); err != nil {
+		return nil, fmt.Errorf("error RewriteConstants: %w", err)
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	opts := ebpf.CollectionOptions{}
+
+	p := &pidIter{}
+	if err := spec.LoadAndAssign(&p.objs, &opts); err != nil {
+		return nil, fmt.Errorf("failed to load ebpf program: %w", err)
+	}
+
+	iter, err := link.AttachIter(link.IterOptions{
+		Program: p.objs.GadgetEbpftop,
+		Map:     nil,
+	})
+	if err != nil {
+		p.objs.Close()
```

Comment:
This `p.objs.Close()` would be easy to forget if we add more `if err != nil` sections below. We cannot just use `defer p.objs.Close()` because we don't want to do it in the success path.

I've started to use the following code pattern to solve this error management issue in various gadgets. In this way, the `defer p.objs.Close()` would not be forgotten if we add more sections below.

* https://github.com/kinvolk/inspektor-gadget/blob/04c2a3664d830cfc39c294f9a37b1b73159de008/pkg/gadgets/dns/tracer/tracer.go#L93-L109
* https://github.com/kinvolk/inspektor-gadget/blob/04c2a3664d830cfc39c294f9a37b1b73159de008/pkg/gadgets/network-graph/tracer/tracer.go#L109-L117
* https://github.com/kinvolk/inspektor-gadget/blob/04c2a3664d830cfc39c294f9a37b1b73159de008/pkg/gadgets/snisnoop/tracer/tracer.go#L97-L109